skimage/feature/texture.py

```diff
@@ -97,11 +101,27 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
     assert_nD(distances, 1, 'distances')
     assert_nD(angles, 1, 'angles')
 
-    assert levels <= 256
     image = np.ascontiguousarray(image)
     assert image.min() >= 0
-    assert image.max() < levels
-    image = image.astype(np.uint8)
+
+    image_max = image.max()
+
+    if np.issubdtype(image.dtype, np.float):
+        raise ValueError("Float images are not supported by greycomatrix. "
+                         "The image needs to be cast to an unsigned integer type.")
+    
+    # for 16 bit images (or larger), levels must be set.
+    if image.dtype != np.uint8:
+        if levels is None:
+            raise ValueError("The levels argument is required for data types other than uint8. "
+                             "The resulting matrix will be at least levels ** 2 in size.")
+                             
+        
```

Comment:
Now that we have decided to support signed integer types, we might want to add another check:

```python
if image.dtype in (np.int8, np.int16, np.int32, np.int64) and np.any(image < 0):
    raise ValueError("greycomatrix is undefined for images with negative values.")
```